File: src/main/java/com/github/jamesgallicchio/stravum/HashUtils.java
package com.github.jamesgallicchio.stravum;

import java.math.BigInteger;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;

public class HashUtils {

    private static final MessageDigest HASHER;

    static {
        MessageDigest md = null;
        try {
            md = MessageDigest.getInstance("SHA-256");
        } catch (NoSuchAlgorithmException e) {
            e.printStackTrace();
        }
        HASHER = md;
    }

    public static byte[] hexToBytes(String s) {
        int len = s.length();
        byte[] data = new byte[len / 2];
        for (int i = 0; i < len; i+= 2) {
            data[i / 2] = (byte) ((Character.digit(s.charAt(i), 16) << 4)
                                 + Character.digit(s.charAt(i+1), 16));
        }
        return data;
    }

    public static String bytesToHex(byte[] d) {
        int lim = d.length * 2;
        StringBuilder sb = new StringBuilder(lim);
        for (int i = 0; i < lim; i += 2) {
            sb.append(Character.forDigit(d[i / 2] >> 4 & 0xF, 16));
            sb.append(Character.forDigit(d[i / 2] & 0xF,  16));
        }
        return sb.toString();
    }

    public static byte[] bytewiseReverse(byte[] b) {
        byte[] d = new byte[b.length];
        for (int i = 0; i < b.length; i++) {
            d[b.length-1-i] = b[i];
        }
        return d;
    }

    public static byte[] sha256(byte[] data, byte[]... moreData) {
        HASHER.update(data);
        for (byte[] b : moreData) {
            HASHER.update(b);
        }
        return HASHER.digest();
    }

    public static byte[] doubleSHA256(byte[] data, byte[]... moreData) {
        return sha256(sha256(data, moreData));
    }

    public static byte[] buildBlockHeaderWithoutNonce(byte[] version, byte[] lastHash, byte[] merkleRoot, int ntime, byte[] nbits) {
        byte[] header = new byte[76];
        copyReversed(version, header, 0);
        copyReversed(lastHash, header, 4);
        copyReversed(merkleRoot, header, 36);
        header[68] = (byte) (ntime);
        header[69] = (byte) (ntime >> 8);
        header[70] = (byte) (ntime >> 16);
        header[71] = (byte) (ntime >> 24);
        copyReversed(nbits, header, 72);
        return header;
    }

    private static void copyReversed(byte[] src, byte[] dest, int destPos) {
        for (int srcPos = src.length - 1; srcPos >= 0 && destPos < dest.length; destPos++, srcPos--) {
            dest[destPos] = src[srcPos];
        }
    }

    public static byte[] blockHash(byte[] header, int nonce) {
        HASHER.update(header);
        HASHER.update((byte) (nonce));
        HASHER.update((byte) (nonce >> 8));
        HASHER.update((byte) (nonce >> 16));
        HASHER.update((byte) (nonce >> 24));
        //HASHER.update(HEADERPADDING);
        return HASHER.digest(HASHER.digest());
    }

    public static final BigInteger TARGET_DIFFICULTY_ONE = new BigInteger("00000000ffff0000000000000000000000000000000000000000000000000000", 16);

    public static byte[] getTargetForDifficulty(int d) {
        return TARGET_DIFFICULTY_ONE.divide(BigInteger.valueOf(d)).toByteArray();
    }

    public static boolean hashLessThanTarget(byte[] hash, byte[] target) {
        if (hash.length != target.length) {
            throw new IllegalArgumentException("Hash and target not same length!");
        }

        for (int i = 0; i < hash.length; i++) {
            if (hash[i] < target[i]) return true;
            else if (hash[i] > target[i]) return false;
        }
        return false;
    }
}

File: src/main/scala/com/github/jamesgallicchio/stravum/HashUtils.scala
package com.github.jamesgallicchio.stravum

import java.math.BigInteger
import java.nio.ByteBuffer
import java.security.MessageDigest

import scala.annotation.tailrec
import scala.collection.mutable

object HashUtils {
  private def newHasher = MessageDigest.getInstance("SHA-256")
  private var hashers = List[MessageDigest]()
  def hasherLock[T](f: MessageDigest => T) = {
    val hasher = hashers.synchronized(hashers match {
      case Nil => newHasher
      case head :: tail => hashers = tail; head
    })
    val res = f(hasher)
    hashers.synchronized(hashers = hasher :: hashers)
    res
  }

  type Hash = Array[Byte]

  /* ------ STRING HEX STUFF ------ */
  def hex(bytes: ByteBuffer): String = hex(bytes.array())
  def hex(bytes: Array[Byte]): String = bytes.foldLeft("") { (s, b) => s + Integer.toHexString(15 & (b >> 4)) + Integer.toHexString(15 & b)}

  implicit class BytesToHex(val bytes: Array[Byte]) extends AnyVal {
    def hex: String = HashUtils.hex(bytes)
  }

  implicit class HexConversions(val hex: String) extends AnyVal {
    def unhex: Array[Byte] = hex.grouped(2).map(b =>
        ((b.charAt(0).asDigit << 4) |
          (if (b.length < 2) 0 else b.charAt(1).asDigit))
        .asInstanceOf[Byte]
      ).toArray
    def intUnhex: Int = Integer.parseUnsignedInt(hex, 16)
  }

  /* ------ HASH STUFF ------ */
  implicit class ByteArrayHasher(val bytes: Array[Byte]) extends AnyVal {
    def hash: Hash = hasherLock(_.digest(bytes))
    def hash2: Hash = hasherLock{ h => h.digest(h.digest(bytes)) }
  }

  def hash(data: Array[Byte]*): Hash = hasherLock{ h =>
    data.foreach(h.update)
    h.digest
  }
  def hash2(data: Array[Byte]*): Hash = hasherLock{ h =>
    data.foreach(h.update)
    h.digest(h.digest)
  }

  /* ------ HASH TARGET STUFF ------ */
  private val target1 = new BigInteger("00000000ffff0000000000000000000000000000000000000000000000000000".unhex)
  private val targets = new mutable.HashMap[Long, Array[Byte]]() // cache of precalculated values

  def targetForDiff(diff: Long): Array[Byte] = targets.getOrElseUpdate(diff, {
    val newTarget = target1.divide(BigInteger.valueOf(diff)).toByteArray
    val arr = new Array[Byte](32)
    System.arraycopy(newTarget, 0, arr, 32 - newTarget.length, newTarget.length)
    arr
  })

  def underTarget(hash: Hash, target: Array[Byte]): Boolean = {
    @tailrec
    def rec(h: Hash, hi: Int, t: Hash, ti: Int): Boolean =
      if (hi < 0 || ti >= t.length) false // Check indexes in range
      else Integer.compareUnsigned(h(hi), t(ti)) match {
        case x if x > 0 => false
        case x if x < 0 => true
        case _ => rec(h, hi - 1, t, ti + 1) // Check if equal => keep comparing
      }

    rec(hash, hash.length - 1, target, 0)
  }
}